output/shared/chunkfactory.go

```diff
@@ -0,0 +1,126 @@
+package shared
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/relex/gotils/logger"
+	"github.com/relex/slog-agent/base"
+	"github.com/relex/slog-agent/util"
+)
+
+type InitCompessorFunc func(log logger.Logger, w io.Writer) io.WriteCloser
+
+type chunkEncoder interface {
+	EncodeChunk(chunk *BasicChunk) ([]byte, error)
+}
+
+type BasicChunkFactory struct {
+	log               logger.Logger
+	reusedChunkBuffer *bytes.Buffer // since chunks are created consecutively, it's possible for them to share the data buffer
+	initCompessor     InitCompessorFunc
+	idGenerator       *chunkIDGenerator
+	chunkEncoder      chunkEncoder
+}
+
+type BasicChunk struct {
+	id           string
+	numRecords   int
+	numBytes     int
+	compressor   io.WriteCloser // could be something like a gzip.Writer or nil to disable compression
+	reusedBuffer *bytes.Buffer  // an actual buffer that compressor writes to
```

Comment:
Yeah, but where are we going to get the actual compressed bytes from, after we're done writing the data? Since the compressor here is just a writer, we need to access its underlying buffer somehow. We could isolate the writer and the buffer in a separate abstraction but I don't really see much need for that.

Reply:
You're right. It cannot be separated.

Another problem: the passing of `Bytes` is rather confusing and dangerous - it's only meaningful inside `FinalizeChunk()` to messageEncoder and if accessed outside it'd return completely wrong data (and neither nil nor empty) due to the buffer being reused.

I'd perhaps pass properties as parameters to messageEncoder or a separate struct created and only used there.

The name `BasicChunk` may be better named as something like "IntermediateChunk".

Reply:
Yeah, I guess decoupling `chunkEncoder` from the chunk itself would be a sound idea, fixed.